In [ ]:
import pandas as pd

# Financial Analysis Summary

This analysis examines key financial metrics from the 10-K filings of Microsoft, Tesla, and Apple for the fiscal years 2021, 2022, and 2023. We calculated year-over-year percentage changes for Total Revenue, Net Income, Total Assets, Total Liabilities, and Cash Flow from Operating Activities to identify trends and shifts in company performance.


In [ ]:
df = pd.read_csv('/content/10K_data.csv')
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0  Microsoft         2021         198270       72738        364840   
1  Microsoft         2022         211915       72361        411976   
2  Microsoft         2023         245122       88136        512163   
3      Tesla         2021          53823        5644         62131   
4      Tesla         2022          81462       12587         82338   
5      Tesla         2023          96773       14974        106618   
6      Apple         2021         365817       94680        351002   
7      Apple         2022         394328       99803        352755   
8      Apple         2023         383285       96995        352583   

   Total Liabilities  Cash Flow from Operating Activities  
0             198298                                89035  
1             205753                                87582  
2             243686                               118548  
3              30548                                11497  
4              36440                                14724  
5              43009                                13256  
6             287912                               104038  
7             302083                               122151  
8             290437                               110543

In [ ]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100
df['Assets Growth (%)'] = df.groupby(['Company'])['Total Assets'].pct_change() * 100
df['Liabilities Growth (%)'] = df.groupby(['Company'])['Total Liabilities'].pct_change() * 100
df['Cash Flow Growth (%)'] = df.groupby(['Company'])['Cash Flow from Operating Activities'].pct_change() * 100
df.fillna(0, inplace=True)
df

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0  Microsoft         2021         198270       72738        364840   
1  Microsoft         2022         211915       72361        411976   
2  Microsoft         2023         245122       88136        512163   
3      Tesla         2021          53823        5644         62131   
4      Tesla         2022          81462       12587         82338   
5      Tesla         2023          96773       14974        106618   
6      Apple         2021         365817       94680        351002   
7      Apple         2022         394328       99803        352755   
8      Apple         2023         383285       96995        352583   

   Total Liabilities  Cash Flow from Operating Activities  Revenue Growth (%)  \
0             198298                                89035            0.000000   
1             205753                                87582            6.882030   
2             243686                               118548           15.669962   
3              30548                                11497            0.000000   
4              36440                                14724           51.351653   
5              43009                                13256           18.795267   
6             287912                               104038            0.000000   
7             302083                               122151            7.793788   
8             290437                               110543           -2.800461   

   Net Income Growth (%)  Assets Growth (%)  Liabilities Growth (%)  \
0               0.000000           0.000000                0.000000   
1              -0.518299          12.919636                3.759493   
2              21.800417          24.318650               18.436183   
3               0.000000           0.000000                0.000000   
4             123.015592          32.523217               19.287678   
5              18.964010          29.488207               18.026894   
6               0.000000           0.000000                0.000000   
7               5.410858           0.499427                4.921990   
8              -2.813543          -0.048759               -3.855232   

   Cash Flow Growth (%)  
0              0.000000  
1             -1.631942  
2             35.356580  
3              0.000000  
4             28.068192  
5             -9.970117  
6              0.000000  
7             17.409985  
8             -9.502992

In [ ]:
summary = df.groupby('Company').agg({
    'Revenue Growth (%)': 'mean',
    'Net Income Growth (%)': 'mean',
    'Assets Growth (%)': 'mean',
    'Liabilities Growth (%)': 'mean',
    'Cash Flow Growth (%)': 'mean'
}).reset_index()
print("\nYear-over-Year Average Growth Rates (%):")
print(summary)


Year-over-Year Average Growth Rates (%):
     Company  Revenue Growth (%)  Net Income Growth (%)  Assets Growth (%)  \
0      Apple            1.664442               0.865772           0.150223   
1  Microsoft            7.517331               7.094040          12.412762   
2      Tesla           23.382306              47.326534          20.670475   

   Liabilities Growth (%)  Cash Flow Growth (%)  
0                0.355586              2.635664  
1                7.398559             11.241546  
2               12.438191              6.032692  


## Microsoft: Consistent Performance

### Revenue Growth
Microsoft experienced average revenue growth of **7.52%**. This indicates steady and healthy growth, driven by strong product demand and the company’s continued success in expanding its cloud computing and enterprise services.

### Net Income Growth
Microsoft’s net income growth averaged **7.09%**, reflecting solid profitability improvements over the period. This shows that Microsoft has been able to translate revenue growth into profit efficiently, likely through its high-margin cloud and software services.

### Asset and Liability Growth
Microsoft’s assets grew at a healthy rate of **12.41%**, while liabilities grew at a similar pace of **7.40%**. This balance between asset and liability growth suggests that Microsoft has been expanding its operations and investing in new ventures, while maintaining a controlled level of debt.

### Cash Flow Growth
The company saw average cash flow growth of **11.24%**, indicating that its core operations are generating strong cash flows, likely driven by increased efficiency and high demand for its services. This is a positive indicator for Microsoft’s financial health.


## Apple: Modest Growth

### Revenue Growth
Apple's revenue growth over the past three years averaged **1.66%**, indicating relatively stable but modest expansion. This slow growth suggests that Apple may be facing market saturation or slowing demand for its products, making it more difficult to achieve the significant revenue growth it experienced in prior years.

### Net Income Growth
Apple’s net income growth averaged **0.87%**, reflecting very limited improvement in profitability. This aligns with the slight revenue growth, suggesting that cost control and profitability measures haven’t significantly boosted the company’s bottom line.

### Asset and Liability Growth
The growth in assets for Apple was minimal at **0.15%**, while liabilities grew slightly more at **0.36%**. These low growth rates indicate that Apple has maintained a relatively stable balance sheet, with little change in its overall assets or liabilities.

### Cash Flow Growth
Apple’s cash flow from operating activities showed an average growth of **2.64%**, which is modest but indicates that the company has managed to slightly improve its cash generation from core operations despite slow revenue and net income growth.


## Tesla: Rapid Expansion

### Revenue Growth
Tesla displayed the most aggressive revenue growth, averaging **23.38%** year-over-year. This high growth rate is indicative of Tesla’s rapid market expansion and growing demand for electric vehicles, as well as its ability to scale production.

### Net Income Growth
Tesla’s net income grew at a remarkable average rate of **47.33%**, reflecting the company’s ability to significantly improve profitability. Tesla has clearly been able to leverage its revenue growth to achieve outsized profit increases, likely due to improved production efficiencies and cost management.

### Asset and Liability Growth
Tesla’s total assets grew at a strong rate of **20.67%**, while liabilities grew at an even higher rate of **12.44%**. This shows that Tesla has been heavily investing in expanding its production capacity and operations, although it has also been taking on more liabilities to finance this growth.

### Cash Flow Growth
Tesla’s cash flow from operating activities grew by an average of **6.03%**, indicating that while the company is generating more cash, its rapid growth may be consuming a significant portion of its cash flows. Tesla’s relatively lower cash flow growth, compared to its revenue and income growth, may suggest the need for continued investment in operations.


## Conclusion

- **Apple** has demonstrated modest growth in both revenue and profitability, indicating a more mature phase in its business lifecycle with limited room for explosive growth. The slight increase in cash flow suggests that Apple’s core operations remain strong, though the company may need new growth strategies to re-accelerate.
  
- **Microsoft** has shown consistent, steady growth across all metrics. Its balanced asset and liability expansion, coupled with strong cash flow generation, suggests that Microsoft is in a stable and healthy financial position, with solid operational efficiency driving both revenue and profitability growth.

- **Tesla** is in a rapid expansion phase, with very high revenue and net income growth rates. However, Tesla’s liability growth is also substantial, reflecting its aggressive investments in capacity expansion. While cash flow is growing, the company may face challenges in sustaining this level of growth without further financial discipline.

Overall, each company demonstrates unique growth characteristics: Apple as a mature, stable company; Microsoft with steady and efficient expansion; and Tesla with rapid growth but potentially higher risk due to its reliance on external financing.
